### PR TITLE
UI: Fix `viewMode` parameter handling

### PR DESCRIPTION
**Pull request by ndelangen**
Issue: https://github.com/storybookjs/storybook/issues/10233

## What I did
Made a change in lib/api/modules/stories so when navigating to a story, the parameter is read and preferred over the current state

WDYT @jonhuteau ?

### CODE REVIEW

**Comment by jonhuteau** [comment]
Looks good to me. So it will be possible to define `viewMode` in these cases ?

```js
// my-component.stories.js

playground.story = {
  parameters: {
    viewMode: 'story',
  },
}
```

```
// my-component.stories.mdx

<Meta
  title='Packages/MyComponent'
  component={MyComponent}
  parameters={{ viewMode: 'docs' }}
/>
```

File: lib/api/src/modules/stories.ts [hunk 1]
```diff
@@ -239,14 +239,13 @@ export const init: ModuleFn = ({
...
 
-      const viewMode = viewModeFromArgs || viewModeFromState;
-
       const hash = ref ? refs[ref].stories : storiesHash;
 
       if (!story) {
         const s = hash[kindOrId] || hash[sanitize(kindOrId)];
         // eslint-disable-next-line no-nested-ternary
         const id = s ? (s.children ? s.children[0] : s.id) : kindOrId;
+        const viewMode = viewModeFromArgs || s ? s.parameters.viewMode : viewModeFromState;
```
Comment (shilman) [inline lib/api/src/modules/stories.ts:248]: what if viewMode is not set in `s.parameters`?

**Review by shilman — CHANGES_REQUESTED** [review]
Looks good conceptually, but maybe buggy?

File: lib/api/src/modules/stories.ts (hunk 1 above)
Comment (ndelangen) [reply lib/api/src/modules/stories.ts:248]: good call, will adjust

File: lib/api/src/lib/stories.ts [hunk 2]
```diff
@@ -58,10 +59,11 @@ export interface Story {
       hierarchyRootSeparator?: RegExp;
       hierarchySeparator?: RegExp;
       showRoots?: boolean;
-      [k: string]: any;
+      [optionName: string]: any;
     };
     docsOnly?: boolean;
-    [k: string]: any;
+    viewMode?: boolean;
```
Comment (ndelangen) [inline lib/api/src/lib/stories.ts:65]: should be ViewMode

Comment (shilman) [reply lib/api/src/lib/stories.ts]: why?

**Comment by ndelangen** [comment]
Seems to work to me:
https://deploy-preview-10292--storybookjs.netlify.app/official-storybook/?path=/docs/addons-docs-buttongroup--basic

Try navigating to a different story, click canvas, then navigate to that story again, it switching you to viewMode=docs

**Pull request merged**